analysis/src/Hint.ml

```diff
@@ -119,4 +119,58 @@ let inlay ~path ~pos ~maxLength ~debug =
                | Some value ->
                  if String.length label > value then None else Some result
                | None -> Some result)
-             | None -> None)))
\ No newline at end of file
+             | None -> None)))
+
+let codeLens ~path ~debug =
+  let lenses = ref [] in
+  let push loc =
+    let range = Utils.cmtLocToRange loc in
+    lenses := range :: !lenses
+  in
+  (* Code lenses are only emitted for functions right now. So look for value bindings that are functions,
+     and use the loc of the value binding itself so we can look up the full function type for our code lens. *)
+  let value_binding (iterator : Ast_iterator.iterator)
+      (vb : Parsetree.value_binding) =
+    (match vb with
+    | {
+     pvb_pat = {ppat_desc = Ppat_var _; ppat_loc};
+     pvb_expr = {pexp_desc = Pexp_fun _};
+    } ->
+      push ppat_loc
+    | _ -> ());
+    Ast_iterator.default_iterator.value_binding iterator vb
+  in
+  let iterator = {Ast_iterator.default_iterator with value_binding} in
+  (if Filename.check_suffix path ".res" then
+   let parser =
+     Res_driver.parsingEngine.parseImplementation ~forPrinter:false
+   in
+   let {Res_driver.parsetree = structure} = parser ~filename:path in
+   iterator.structure iterator structure |> ignore);
+  !lenses
+  |> List.filter_map (fun (range : Protocol.range) ->
+         match Cmt.fullFromPath ~path with
+         | None -> None
+         | Some full -> (
+           match
+             References.getLocItem ~full
+               ~pos:(range.start.line, range.start.character + 1)
+               ~debug
+           with
+           | Some {locType = Typed (_, typeExpr, _)} ->
```

Comment:
I have zero knowledge about analysis opt-in in the vscode. Please forgive me about maybe nonsense question.

I always want to have a feature of type checking on the fly, I mean `type check as you type`. This line of code hits me. The analysis can check the type? I mean it has an opt-in type checker already?

Reply:
> I have zero knowledge about analysis opt-in in the vscode. Please forgive me about maybe nonsense question.
> 
> I always want to have a feature of type checking on the fly, I mean `type check as you type`. This line of code hits me. The analysis can check the type? I mean it has an opt-in type checker already?

Quick background:
- When the editor integration starts running, the compiler has already run (there are exceptions for unsaved code, but let's ignore this)
- The compiler artifacts contain type information.

That type information is used by the analysis as-you-type, as long as the type information was already present before you start typing (i.e. last time the file was saved).